language-plutus-core/src/Language/PlutusCore/Analysis/Definitions.hs

```diff
@@ -0,0 +1,167 @@
+{-# LANGUAGE LambdaCase #-}
+-- | Definition analysis for Plutus Core.
+module Language.PlutusCore.Analysis.Definitions (UniqueInfos, ScopeType, termDefs, typeDefs, runTermDefs, runTypeDefs) where
+
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Type
+
+import           Data.Functor.Foldable
+
+import           Control.Lens              hiding (use, uses)
+import           Control.Monad.Except
+import           Control.Monad.State
+import           Control.Monad.Writer
+
+import           Data.Coerce
+import           Data.Foldable
+import qualified Data.IntMap               as IM
+import qualified Data.Set                  as Set
+
+{- Note [Unique usage errors]
+The definitions analysis can find a number of problems with usage of uniques, however
+clients may not always want to throw an error on all of them, hence we simply return
+them all and allow the client to chose if they want to throw some of them.
+-}
+
+-- | Information about a unique, a pair of a definition if we have one and a set of uses.
+type UniqueInfo a = (Maybe (ScopedLoc a), Set.Set (ScopedLoc a))
+type UniqueInfos a = IM.IntMap (UniqueInfo a)
+
+data ScopedLoc a = ScopedLoc ScopeType a deriving (Eq, Ord)
+
+-- | Tag for distinguishing between whether we are talking about the term scope
+-- for variables or the type scope for variables.
+data ScopeType = TermScope | TypeScope deriving (Eq, Ord)
+
+lookupDef
+    :: (Ord a,
+        MonadState (UniqueInfos a) m)
+    => Unique
+    -> m (UniqueInfo a)
+lookupDef u = do
+    previousDef <- gets $ IM.lookup (unUnique u)
+    case previousDef of
+        Just d -> pure d
+        Nothing -> do
+            let empty = (Nothing, mempty)
+            modify $ IM.insert (unUnique u) empty
+            pure empty
+
+addDef
+    :: (Ord a,
+        HasUnique n unique,
+        MonadState (UniqueInfos a) m,
+        MonadWriter [UniqueError a] m)
```

Comment:
It's rather unfortunate that none of the variants of `Validation` ([1](http://hackage.haskell.org/package/validation-1/docs/Data-Validation.html), [2](http://hackage.haskell.org/package/either-5.0.1/docs/Data-Either-Validation.html)) have `MonadValidation`.